File: src/java/com/dgm/api/src/TermType.java
/*
 * Copyright (C) 2017 Benoit CAYLA (benoit@famillecayla.fr)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.dgm.api.src;

import com.joy.api.JoyApiRequestParameter;
import com.joy.api.beans.JoyJsonPOSTReturn;
import static com.joy.api.beans.JoyJsonPOSTReturn.JoyEnumPOSTUpSertCodes.*;
import com.joy.api.beans.JoyJsonVector;
import com.joy.api.utils.RESTEntityCommon;
import com.joy.bo.BOEntityReadWrite;
import com.joy.bo.IEntity;
import com.joy.common.parameters.JoyParameter;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.List;
import java.util.logging.Level;

/**
 *
 * @author Benoit Cayla (benoit@famillecayla.fr)
 */
public class TermType extends RESTEntityCommon {

    /**
     * return the GIO_PK value (in URL params)
     * @return GIO_PK
     */
    private int getId() {
    	try {
            return Integer.valueOf(this.getCurrentRequest().getParameter("GIO_PK").getValue()); 
        } catch (NumberFormatException e) {}
        return 0;
    }

    /**
     * Delete an row in SRC_TERMTYPE table
     * Filter with the ID if given in the URL
     * DELETE http://[...]/api/termtype?GIO_PK=X
     * @return json content
     */
    @Override
    public String restDelete(JoyJsonPOSTReturn retDELETE) {
        int uid = getId();
        if (uid != 0) {
            BOEntityReadWrite Entity = (BOEntityReadWrite)this.getBOFactory().getEntity("SRC_TERMTYPE");
            Entity.reset();
            Entity.field("GIO_PK").setKeyValue(uid);
            int intCountRecAffected = Entity.delete();
            if (intCountRecAffected != 1) {
                retDELETE.setStatusKo();
                retDELETE.addMessage("The item has not been deleted successfully");
            } else retDELETE.setStatusOk();
        }
        return this.getStatusOk();
    }

    /**
     * Check if the uid already exists in the table
     * @param uid
     * @param retPOST
     * @return 
     */
    private boolean checkExisting(String termname, JoyJsonPOSTReturn retPOST) {
        try {
            IEntity Entity = (BOEntityReadWrite)this.getBOFactory().getEntity("SRC_TERMTYPE");
            Entity.field("GIO_TERMTYPE_NAME").setKeyValue(termname);
            boolean exists = Entity.hasRecord();
            if (exists) {
                retPOST.setStatusKo();
                retPOST.addMessage("The term's name you tried to insert already exists");
            }
            return exists;
            
        } catch (Exception e) {
            return true;
        }
    }
    
    /**
     * Add or update a new record (upsert mode)
     * POST http://[...]/api/termtype
     *      * execute an insert if param[GIO_PK] == 0
     *      * execute an update else
     * fields in the parameters POST request
     * @return 
     */
    @Override
    public String restPost(JoyJsonPOSTReturn retPOST) {
        int uid = getId();
        int intCountRecAffected;
        
        try {
            IEntity Entity = (BOEntityReadWrite)this.getBOFactory().getEntity("SRC_TERMTYPE");
            Entity.field("GIO_TERMTYPE_NAME").setValue(this.getCurrentRequest().getParameter("GIO_TERMTYPE_NAME").getValue());
            Entity.field("GIO_ICON_PATHNAME").setValue(this.getCurrentRequest().getParameter("GIO_ICON_PATHNAME").getValue());
            if (uid == 0) {
                // Checks if the record already exists
                intCountRecAffected = 0;
                if (!checkExisting(this.getCurrentRequest().getParameter("GIO_TERMTYPE_NAME").getValue(), retPOST)) {
                    // Insert a new record
                    retPOST.setUpdateType(insert);
                    int newid = Entity.getNewIDForField("GIO_PK");
                    Entity.field("GIO_PK").setValue(newid);
                    intCountRecAffected = Entity.insert();
                    if (intCountRecAffected != 1) {
                        retPOST.setStatusKo();
                        retPOST.addMessage("The item has not been inserted successfully");
                    } else retPOST.setStatusOk();
                }
            } else {
                // update an existing record
                retPOST.setUpdateType(update);
                Entity.field("GIO_PK").setKeyValue(uid);
                intCountRecAffected = Entity.update();
                if (intCountRecAffected != 1) {
                    retPOST.setStatusKo();
                    retPOST.addMessage("The item has not been updated successfully");
                } else retPOST.setStatusOk();
            } 
            retPOST.setNbRowsAffected(intCountRecAffected);

        } catch (Exception e) {
            getLog().log(Level.SEVERE, "Error while creating a new Term type, {0}", e.toString());
            this.getStatusAlreadyExist();
        }
        return this.getStatusOk();
    }

    /**
     * Return the SRC_TERMTYPE table content
     * Filter with the ID if given in the URL
     * GET http://[...]/api/termtype (full table)
     * GET http://[...]/api/termtype?GIO_PK=X (only record with id)
     * @return json content
     */
    @Override
    public String restGet() {
        try {
            // Table data
            IEntity Entity = (BOEntityReadWrite)this.getBOFactory().getEntity("SRC_TERMTYPE");
            JoyApiRequestParameter param = this.getCurrentRequest().getParameter("GIO_PK");
            if (param == null) { // Display list
                // return all the table content
                this.addMatrix("SRC_TERMTYPE", Entity);
                
                // Combobox term's icons
                List<JoyParameter> icons = this.getState().getAppParameters().getParameter("TermTypeIcons").getList();
                JoyJsonVector  columns = new JoyJsonVector();
                for (JoyParameter icon : icons) {
                    columns.addItem(icon.getValue().toString(), icon.getValue().toString());
                }
                this.addVector("GIO_ICON_PATHNAME", columns);

                // Combobox term type (list of all glossaries)
                JoyJsonVector glossaries = new JoyJsonVector();
                IEntity entity = this.getBOFactory().getEntity("DIM_GLOSSARY");
                entity.addSort("GLO_NAME");
                ResultSet rs = entity.select();
                while (rs.next()) {
                    glossaries.addItem(rs.getString("GLO_NAME"), rs.getString("GLO_NAME"));
                }
                this.getBOFactory().closeResultSet(rs);
                this.addVector("GIO_TERMTYPE_NAME", glossaries);
                
            } else { // Display 1 item only
                Entity.field("GIO_PK").setKeyValue(param.getValue());
                this.addSingle(Entity);
            }
            return this.getData().toString();
            
        } catch (SQLException ex) {
            this.getLog().severe(ex.toString());
            return super.getStatusNoContent();
        }
    }
    
}

File: web/app/configuration/termtype/termtype.js
/* 
 * Copyright (C) 2017 Benoit CAYLA (benoit@famillecayla.fr)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

function cb_update_return(content) {
    if (content.status === 'OK') {
        $("#myModal").modal('hide');
        refreshList();
    } else {
        // display error message
        $("#errormessages").show();
        $("#errormessages").html($$.getFormattedErrorMessages(content));
    }
}

function cb_delete_return(content) {
    if (content.status != 'OK') 
        bootbox.alert($$.getFormattedErrorMessages(content));
    refreshList();
}

function evt_popupSave () {
    var postParams = {
        "GIO_PK" : $("#GIO_PK").html(),
        "GIO_TERMTYPE_NAME" : $("#gio_termtype_name").val(),
        "GIO_ICON_PATHNAME" : $("#gio_icon_pathname").val()
    };
    // Update or create the item
    $$.ajax("POST", cb_update_return, $$.getAPICall("srctermtype"), postParams);
}

function cb_displayBoxWithData(itemdata) {
    init_popupControl();
    // Identifier
    $("#GIO_PK").html($$.getData(itemdata.single, "gio_pk"));
    // Combo term's type names
    $("#gio_termtype_name").val($$.getData(itemdata.single, "gio_termtype_name")).trigger("change");
    $("#gio_termtype_name").prop("disabled", true); // disable the combo !
    // Combo path names
    $("#gio_icon_pathname").val($$.getData(itemdata.single, "gio_icon_pathname")).trigger("change");
    // Display modal
    $("#myModal").modal('show');
}

function init_popupControl() {
    $("#errormessages").hide();
    $("#errormessages").html("");
    $('#gio_termtype_name').select2({ placeholder: "Select an Term type", width: '100%'  });
    $('#gio_icon_pathname').select2({ placeholder: "Select an icon", width: '100%'  });
}

function evt_requestInsert() {
    init_popupControl();
    $("#GIO_PK").html("");
    $("#gio_termtype_name").val("").trigger("change");
    $("#gio_termtype_name").prop("disabled", false); // enable the combo !
    $("#gio_icon_pathname").val("").trigger("change");
    // Display modal
    $("#myModal").modal('show');
}

function evt_requestUpdate(id) {
    $$.ajax("GET", cb_displayBoxWithData, $$.getAPICall("srctermtype", { "GIO_PK" : id }));
}

function evt_requestDelete(id) {
    $$.ajax("DELETE", cb_delete_return, $$.getAPICall("srctermtype", { "GIO_PK" : id }));
}

function cb_refreshList(content) {
    var t1 = $('#myList').DataTable();
    var tableContent = $$.getData(content.matrix, "SRC_TERMTYPE");
    t1.clear();
    $("#pleasewait").html('');
    for (i=0; i < tableContent.rowcount; i++) {
        var id = tableContent.rows[i].items[0].value;
        t1.row.add( [
            id,
            "<A href='#' onClick='evt_requestUpdate(" + id + ")'>" + tableContent.rows[i].items[1].value + "</a>",
            tableContent.rows[i].items[2].value,
            "<A href='#' onClick='evt_requestDelete(" + id + ")'>[delete]</a>"
        ] ).draw( false );
    }
    
    // Init cbo term's from parameter data
    var cboObject = document.getElementById("gio_icon_pathname");
    for (var i=0; i < $$.getContext().parameters.TermTypeIcons.length; i++) {
        var myoption = document.createElement("option");
        myoption.text = $$.getContext().parameters.TermTypeIcons[i].value;
        myoption.value = $$.getContext().parameters.TermTypeIcons[i].value;
        cboObject.add(myoption, null);
    }
}

function cb_initComboboxTermsList(content) {
    $$.fillComboboxFromJoyEntity("gio_termtype_name", content, "glo_name", "glo_name"); 
}

function refreshList() {
    $$.ajax("GET", cb_refreshList, $$.getAPICall('srctermtype'));
}

$$.form_afterLoad = function() {
    init_menus("config");
    $('#myList').DataTable({ responsive: true });
    $$.ajax("GET", cb_initComboboxTermsList, $$.getAPICall('entity/DIM_GLOSSARY'));
    refreshList();
}

$$.init();